# Milestone 3 Assignment

## Background

    The capstone project focuses on diaper manufacturing quality. In the Lesson 01 assignment, you discovered how the diaper manufacturing process works. Generally, to ensure or predict quality, a diaper manufacturer need s to monitor every step of the manufacturing process with sensors such as heat sensors, glue sensors, glue level, etc.

    For this capstone project, we will use the SECOM manufacturing Data Set from the UCI Machine Learning Repository. The set is originally for semiconductor manufacturing, but in our case, we will assume that it is for the diaper manufacturing process.

    The dataset consists of two files:

    (1) a dataset file SECOM containing 1567 examples, each with 591 features, presented in a 1567 x 591 matrix
    (2) a labels file listing the classifications and date time stamp for each example
    
    Reference
    Dua, D. and Karra Taniskidou, E. (2017). UCI Machine Learning Repository [http://archive.ics.uci.edu/ml]. Irvine, CA: University of California, School of Information and Computer Science.

## Instructions

    Create a new notebook and perform each of the following tasks and answer the related questions:

    (1) Build a simple neural networks model
    (2) Build a DNN model
    (3) Build a RNN model
    (4) Summarize your findings with examples.  Explain what the manufacturer should focus on to optimize the diaper manufacturing process.

### (1) Build a simple neural networks model

#### Comments:

    For my simple neural networks model I will be building a Keras model with only one hidden layer.

In [1]:
# Import packages

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from collections import OrderedDict
from matplotlib import pyplot
from sklearn.model_selection import train_test_split
from sklearn import svm, metrics
from sklearn.metrics import accuracy_score
from sklearn import preprocessing

# TensorFlow and tf.keras

import tensorflow as tf
from tensorflow import keras
from keras.datasets import reuters

#import numpy and matplotlib

import numpy as np
from keras.utils import to_categorical

/Users/matt.denko/anaconda3/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.


In [2]:
# Load dataset

url = "https://archive.ics.uci.edu/ml/machine-learning-databases/secom/secom.data"
names = ["feature" + str(x) for x in range(1, 591)]
secom_var = pd.read_csv(url, sep=" ", names=names, na_values = "NaN") 
 
url_l = "https://archive.ics.uci.edu/ml/machine-learning-databases/secom/secom_labels.data"
secom_labels = pd.read_csv(url_l,sep=" ",names = ["classification","date"],parse_dates = ["date"],na_values = "NaN")
feature_names = secom_labels

#Merging the two datasets

data = pd.merge(secom_var, secom_labels,left_index=True,right_index=True)
data.describe()
data.head()
print(data.columns) 
(nrows, ncols) = data.shape

Index(['feature1', 'feature2', 'feature3', 'feature4', 'feature5', 'feature6',
       'feature7', 'feature8', 'feature9', 'feature10',
       ...
       'feature583', 'feature584', 'feature585', 'feature586', 'feature587',
       'feature588', 'feature589', 'feature590', 'classification', 'date'],
      dtype='object', length=592)


In [3]:
# Replacing Nulls

data = data.replace(to_replace= -1, value=float(0))
data = data.replace(to_replace= float('NaN'), value=float(0))
data_null = data.isnull().sum()
print(data_null)
print("There are 0 columns with missing data")

feature1          0
feature2          0
feature3          0
feature4          0
feature5          0
feature6          0
feature7          0
feature8          0
feature9          0
feature10         0
feature11         0
feature12         0
feature13         0
feature14         0
feature15         0
feature16         0
feature17         0
feature18         0
feature19         0
feature20         0
feature21         0
feature22         0
feature23         0
feature24         0
feature25         0
feature26         0
feature27         0
feature28         0
feature29         0
feature30         0
                 ..
feature563        0
feature564        0
feature565        0
feature566        0
feature567        0
feature568        0
feature569        0
feature570        0
feature571        0
feature572        0
feature573        0
feature574        0
feature575        0
feature576        0
feature577        0
feature578        0
feature579        0
feature580        0
feature581        0


In [4]:
# Define the target and features:

target_label = 'classification'
non_feature = 'date'
feature_labels = [x for x in data.columns if x not in [target_label]+ [non_feature]]

# One-hot encode inputs

data_expanded = pd.get_dummies(data, drop_first=True)
print('DataFrame one-hot-expanded shape: {}'.format(data_expanded.shape))

# Get target and original x-matrix

Y = data[target_label]
X = data.as_matrix(columns=feature_labels)

DataFrame one-hot-expanded shape: (1567, 592)


/Users/matt.denko/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:15: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.
  from ipykernel import kernelapp as app


In [5]:
# Split dataset into training set and test set

X_train, X_test, y_train, y_test = train_test_split(X, Y, 
                                  test_size=0.3,random_state=42) # 70% training and 30% test

In [6]:
# Preprocessing Model

model = keras.Sequential([
    keras.layers.Flatten(input_shape=(590,)),
    keras.layers.Dense(1, activation=tf.nn.relu)
])

In [7]:
# Compile Model

model.compile(optimizer='adam', 
              loss='binary_crossentropy',
              metrics=['accuracy'])

# Train Model

model.fit(X_train, y_train, epochs=5)

Epoch 1/5
1096/1096 [==============================] - 0s 227us/step - loss: 4.4015 - acc: 0.7062
Epoch 2/5
1096/1096 [==============================] - 0s 36us/step - loss: 2.8613 - acc: 0.8130
Epoch 3/5
1096/1096 [==============================] - 0s 38us/step - loss: 2.4415 - acc: 0.8412
Epoch 4/5
1096/1096 [==============================] - 0s 40us/step - loss: 1.3354 - acc: 0.9170
Epoch 5/5
1096/1096 [==============================] - 0s 41us/step - loss: 1.2772 - acc: 0.9206


In [8]:
# Evaluating Accuracy Score

test_loss, test_acc = model.evaluate(X_test, y_test)
print('Test accuracy:', test_acc)

471/471 [==============================] - 0s 102us/step
Test accuracy: 0.925690021737619


### (2) Build a DNN model

In [9]:
# 1 - Adding More layers

model = keras.Sequential([
    keras.layers.Flatten(input_shape=(590,)),
    keras.layers.Dense(256, activation=tf.nn.relu),
    keras.layers.Dense(256, activation=tf.nn.relu),
    keras.layers.Dense(256, activation=tf.nn.relu),
    keras.layers.Dense(500, activation=tf.nn.relu),
    keras.layers.Dense(500, activation=tf.nn.relu),
    keras.layers.Dense(1, activation=tf.nn.softmax)
])

model.compile(optimizer='adam', 
              loss='binary_crossentropy',
              metrics=['accuracy'])

model.fit(X_train, y_train, epochs=10)

test_loss, test_acc = model.evaluate(X_test, y_test)

print('Test accuracy:', test_acc)

predictions = model.predict(X_test)

Epoch 1/10
1096/1096 [==============================] - 1s 848us/step - loss: 14.8805 - acc: 0.0666
Epoch 2/10
1096/1096 [==============================] - 0s 241us/step - loss: 14.8805 - acc: 0.0666
Epoch 3/10
1096/1096 [==============================] - 0s 238us/step - loss: 14.8805 - acc: 0.0666
Epoch 4/10
1096/1096 [==============================] - 0s 236us/step - loss: 14.8805 - acc: 0.0666
Epoch 5/10
1096/1096 [==============================] - 0s 230us/step - loss: 14.8805 - acc: 0.0666
Epoch 6/10
1096/1096 [==============================] - 0s 237us/step - loss: 14.8805 - acc: 0.0666
Epoch 7/10
1096/1096 [==============================] - 0s 234us/step - loss: 14.8805 - acc: 0.0666
Epoch 8/10
1096/1096 [==============================] - 0s 255us/step - loss: 14.8805 - acc: 0.0666
Epoch 9/10
1096/1096 [==============================] - 0s 241us/step - loss: 14.8805 - acc: 0.0666
Epoch 10/10
471/471 [==============================] - 0s 202us/step
Test accuracy: 0.06581740980600

### (3) Build a RNN model

In [10]:
# Construct model 3

X_train = np.clip(X_train, 0,9999)
X_test = np.clip(X_test, 0, 9999)


max_review_length = 590 
num_of_words = 10000
embedding_vecor_length = 46
model = keras.models.Sequential()
model.add(keras.layers.Embedding(num_of_words, embedding_vecor_length, input_length=max_review_length))
model.add(keras.layers.LSTM(100))
model.add(keras.layers.Dense(4579, activation='relu'))
model.compile(loss='sparse_categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
print(model.summary())
model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=3, batch_size=64)

test_loss, test_acc = model.evaluate(X_test, y_test)

print('Test accuracy:', test_acc)

predictions = model.predict(X_test)

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding (Embedding)        (None, 590, 46)           460000    
_________________________________________________________________
lstm (LSTM)                  (None, 100)               58800     
_________________________________________________________________
dense_7 (Dense)              (None, 4579)              462479    
Total params: 981,279
Trainable params: 981,279
Non-trainable params: 0
_________________________________________________________________
None


/Users/matt.denko/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/gradients_impl.py:108: UserWarning: Converting sparse IndexedSlices to a dense Tensor of unknown shape. This may consume a large amount of memory.
  "Converting sparse IndexedSlices to a dense Tensor of unknown shape. "


Train on 1096 samples, validate on 471 samples
Epoch 1/3
1096/1096 [==============================] - 18s 17ms/step - loss: 6.7460 - acc: 0.3613 - val_loss: 0.3918 - val_acc: 0.9342
Epoch 2/3
1096/1096 [==============================] - 18s 16ms/step - loss: 0.3198 - acc: 0.9334 - val_loss: 0.3068 - val_acc: 0.9342
Epoch 3/3
471/471 [==============================] - 2s 5ms/step
Test accuracy: 0.9341825907397422


### (4) Summarize your findings with examples.  Explain what the manufacturer should focus on to optimize the diaper manufacturing process.

#### Comments:

    In Milestone 1, I originally ran a model using all 591 features to predict whether or not a quality diaper was produced. This model had very low accuracy. To improve the model, I then used a SMOTE method to handle class imbalance, this improved the accuracy significantly however it was still only .57. I then used recursive feature selection to handle the issue of overfitting. This improved the accuracy to .932. Then in the Milestone 2 assignment, I first ran a decision tree model which produced an accuracy of .932. I then ran a random forest model which produced an accuracy of .934 which is the highest accuracy that I have gottent in predicting whether or not that a quality diaper was produced. For this final Milestone assignment I then ran three different neural networks model. The first model that I ran was a simple neural networks model which was a keras model with a binary crossentropy loss function and one hidden layer. This model produced an accuracy of 81%. The second model that I ran was a similar dnn neural networks model with a binary crossentropy loss function, however I added on additional hidden layers. The addition of hidden layers actually made the accuracy of the model decrease and produced a score of 7%. I then ran a RNN neural networks model using a sparse_categorical_crossentropy loss function which produced the highest accuracy of any of the models with 93.4% accuracy. Based on the results, my reccomendation to the diaper manufacturer would be to not invest in leveraging neural networks to optimize the manufacturing process. The accuracy score of the best neural networks model was comparable to the accuracy score of a random forest model. The neural networks model will be more expensive and harder to act on then a random forest model. 